### ISSUE TITLE
PGFunction identity doesn't take source code into account

### ISSUE THREAD

**Issue by jasiek** [olirice/alembic_utils#2]
Looks like only the signature is taken into account when comparing functions, rather than the body of the function.

This isn't super easy to solve generically across all languages supported by postgres, but I think a good effort could be made on SQL & PL/pgSQL. How about this:
* Retrieve the source for the function,
* Normalise the code (either an existing SQL formatter, or cook something up in python that's good enough)
* Compare the source.

The code will live in `pg_proc.prosrc` for PL/pgSQL, not sure about others.

What do you think?

**Comment by olirice** [comment]
The function definition is used when comparing functions

Here we loop each local definition and look up what migration operation (if any) it needs
https://github.com/olirice/alembic_utils/blob/0b6697ab15e11c1fb3f06c23a5d8b5d0fc4aa4a4/src/alembic_utils/replaceable_entity.py#L353

Here is where the comparison between the function and the existing database entities
https://github.com/olirice/alembic_utils/blob/0b6697ab15e11c1fb3f06c23a5d8b5d0fc4aa4a4/src/alembic_utils/replaceable_entity.py#L189

Any here is where the source definition is retrieved for comparison
https://github.com/olirice/alembic_utils/blob/0b6697ab15e11c1fb3f06c23a5d8b5d0fc4aa4a4/src/alembic_utils/pg_function.py#L106

If you have found an instance where updating the local definition of a PGFunction does not get detected and generate an `update_entity` in the migration, please open an issue

**Comment by jasiek** [comment]
I think the problem is in the chained trims. See below:
```
investor_api=# select prosrc from pg_proc where proname like 'accrue%';
                                                           prosrc
-----------------------------------------------------------------------------------------------------------------------------
 begin                                                                                                                      +
 RETURN QUERY                                                                                                               +
 WITH                                                                                                                       +
 yesterdays_balance AS                                                                                                      +
   (SELECT t1.account_id,                                                                                                   +
(.... truncated ....)
    FROM last_tx_ids l                                                                                                      +
    JOIN yesterdays_balance b ON (l.account_id = b.account_id)                                                              +
    JOIN todays_withdrawals w ON (w.account_id = b.account_id)) AS x                                                        +
 RETURNING transactions.account_id, transactions.amount;                                                                    +
 end                                                                                                                        +

(1 row)
```

Whereas the trim chain will return the following:

```
investor_api=# select trim(trim('\n', trim(trim('\t' from trim('\n' from trim(trim('\n' from trim(prosrc)))))))) from pg_proc where proname like 'accrue%';
 btrim
-------
 \
(1 row)
```

This will collapse all function sources to that string. I think the intention was to use `BTRIM(long_string, characters)` instead which is also available as `TRIM(long_string, characters)`. With this change, the functionality appears to be working.

**Comment by jasiek** [comment]
`TRIM(characters FROM long_string)` would also work. I'll submit a PR in a bit.

**Comment by olirice** [comment]
Fixed with
41af355102f86e4f4f4ac8d256906c1ada309559

trim would have worked too but, but escape characters when manually pattern matching always makes me nervous

It is available on pypi in version 0.2.4

**Comment by jasiek** [comment]
Excellent - thanks.
